get_num_bytes small fixes

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170342
* __->__ #170189

- Instead of using untyped storage(), which will over-estimate offsets into tensor, account for expanded/broadcast dims 
- Handle SymInts by using hints when available
- Remove MIN_ALLOCATE rounding (allocation overhead, not relevant for bandwidth estimation).

eellison committed:
Fix get_num_bytes to estimate accessed bytes, not allocated

- Account for expanded/broadcast dims (stride=0) where only 1 element
  is accessed regardless of logical size
- Handle SymInts by using hints when available
- Remove MIN_ALLOCATE rounding (allocation overhead, not relevant for
  bandwidth estimation)

[ghstack-poisoned]

Skylion007 commented on torch/utils/_runtime_estimation.py:
```diff
@@ -122,9 +114,16 @@ def get_num_bytes(t: torch.Tensor) -> int:
-    mem_consumed = math.ceil(num_bytes / _PYTORCH_MIN_ALLOCATE) * _PYTORCH_MIN_ALLOCATE
-    return mem_consumed
+    real_numel = 1
+    for size, stride in zip(t.shape, t.stride()):
```
Please pass strict=True to zip

eellison replied:
if shape and stride have different lengths we're beyond saving haha

bdhirsh reviewed (approved):
lgtm !

eellison committed:
Update on "get_num_bytes small fixes"


- Instead of using untyped storage(), which will over-estimate offsets into tensor, account for expanded/broadcast dims 
- Handle SymInts by using hints when available
- Remove MIN_ALLOCATE rounding (allocation overhead, not relevant for bandwidth estimation).

[ghstack-poisoned]

eellison commented on torch/utils/_runtime_estimation.py:
```diff
@@ -122,9 +114,16 @@ def get_num_bytes(t: torch.Tensor) -> int:
+    real_numel = 1
+    for size, stride in zip(t.shape, t.stride()):
+        if not has_hint(size) or not has_hint(stride):
+            return 0
```
yea, fair

eellison committed:
Update on "get_num_bytes small fixes"


- Instead of using untyped storage(), which will over-estimate offsets into tensor, account for expanded/broadcast dims 
- Handle SymInts by using hints when available
- Remove MIN_ALLOCATE rounding (allocation overhead, not relevant for bandwidth estimation).

[ghstack-poisoned]
